runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -7114,145 +6829,266 @@ static void genInitArrayHeader(
+
+      TR::Register *zeroInitScratchReg = NULL;
       if (comp->target().is64Bit())
          {
-         scratchReg = cg->allocateRegister();
-         generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, scratchReg, targetReg, cg);
+         zeroInitScratchReg = srm->findOrCreateScratchRegister();
          }
-      else
+
+#ifdef TR_TARGET_64BIT
+      if (sizeReg && doInlineRepStosZeroInit)
          {
-         generateRegInstruction(TR::InstOpCode::PUSHReg, node, targetReg, cg);
+         int32_t repSTOSThresholdAdjustment = 0;
+         int32_t startingAddressAdjustment = 0;
+
+         /**
+          * When initialization will be done inline with 8-byte stores, if the object
+          * header is not a multiple of 8 then adjust the starting address back by
+          * four bytes to ensure the stores are aligned on their natural boundary
+          * and so that initialization does not run past the end of the newly allocated
+          * object.  The REP STOS threshold will also need to be adjusted to account
+          * for these four bytes.
+          */
+         if (headerSizeInBytes % 16 == 12)
+            {
+            repSTOSThresholdAdjustment = -4;
+            startingAddressAdjustment = -4;
+            }
+
+         TR::LabelSymbol *repStosInitLabelSym = generateLabelSymbol(cg);
+         TR::LabelSymbol *mergeInitLabelSym = generateLabelSymbol(cg);
+
+         generateRegImmInstruction(
+            TR::InstOpCode::CMPRegImms(),
+            node,
+            numBytesToZeroInitReg,
+            repStosZeroInitThresholdBytes + repSTOSThresholdAdjustment,
+            cg);
+         generateLabelInstruction(TR::InstOpCode::JG4, node, repStosInitLabelSym, cg);
+
+         // Begin zero initialization after the header, adjusting for header size
+         // if necessary
+         //
+         generateRegMemInstruction(
+            TR::InstOpCode::LEARegMem(),
+            node,
+            segmentReg,
+            generateX86MemoryReference(newObjectAddressReg, headerSizeInBytes + startingAddressAdjustment, cg),
+            cg);
+
+         generateRegRegInstruction(TR::InstOpCode::XOR4RegReg, node, zeroInitScratchReg, zeroInitScratchReg, cg);
+
+         // Generate mainline zero initialization with stores
+         //
+         TR::LabelSymbol *zeroInitLoopLabelSym = generateLabelSymbol(cg);
+         generateLabelInstruction(TR::InstOpCode::label, node, zeroInitLoopLabelSym, cg);
+         generateMemRegInstruction(TR::InstOpCode::S8MemReg, node,
+            generateX86MemoryReference(segmentReg, 0, cg),
+            zeroInitScratchReg, cg);
+         generateRegImmInstruction(TR::InstOpCode::ADD8RegImms, node, segmentReg, 8, cg);
+         generateRegImmInstruction(TR::InstOpCode::SUB8RegImms, node, numBytesToZeroInitReg, 8, cg);
+         generateRegImmInstruction(TR::InstOpCode::CMP8RegImms, node, numBytesToZeroInitReg, 0, cg);
```

Comment:
I think for this particular branch (JG, which does not rely on the carry flag) the CMP to 0 can be eliminated because the relevant flags will have been set up already by the SUB.  I will change it, but I want to re-run the tests (internally) though.

Reply:
Testing this change internally on x86-64 Linux with sanity and openjdk was successful.  I've amended and rebased the commits.